sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py

```diff
@@ -229,10 +235,9 @@ def _initialize_window_of_centroids(
             # The first `window_n_centroids` work items cooperate on loading the
             # values of centroids_half_l2_norm relevant to current window. Each work
             # item loads one single value.
-            if local_work_id < window_n_centroids:
-                half_l2_norm_loading_idx = first_centroid_idx + local_work_id
-                window_of_centroids_half_l2_norms[local_work_id] = (
-                    centroids_half_l2_norm[half_l2_norm_loading_idx]
+            if local_row_idx == zero_as_a_long:
```

Comment:
Are you sure this change is correct? Don't we need a bound check on the value of `local_col_idx`? I would have maybe expected:

```suggestion
            if local_row_idx == zero_as_a_long and local_col_idx < window_n_centroids:
```

assuming that `window_of_centroids_half_l2_norms` is raveled window of logical shape `(1, window_n_centroids)`, but the comment should probably be updated to better explain the interaction between `local_row_idx`, `local_col_idx` and the shape of the window array.

Reply:
Well spotted, I think your suggestion is right. There is a risk to read `centroids_half_l2_norm` array in global memory outside of its boundaries. Not sure why execution on cpu doesn't segfaults :thinking: 

My guess is that the compiler detected that the corresponding write in local memory is not used later on (which is True) and optimized in a way that skipped the risk of segfault.

Reply:
Will update comment accordingly